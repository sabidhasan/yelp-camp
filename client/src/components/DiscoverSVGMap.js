import React from 'react'
// import { Switch, Route } from 'react-router-dom'
import { provinces } from '../helpers/helpers'
import CanadaMapSVG from '../helpers/CanadaMapSVG'

class DiscoverMap extends React.Component {
  constructor() {
    super();
    //state keeps track of currently hovered province
    this.state = {
      hoveredProvince: null
    }
    this.handleClick = this.handleClick.bind(this);
    this.handleHover = this.handleHover.bind(this);
    this.getProvinceName = this.getProvinceName.bind(this);
  }

  componentDidMount() {
    document.title = 'YelpCamp | Discover campgrounds in Canada'
  }

  getProvinceName(e) {
    // if hovered item isnt in svg, return null (could be hovering on parent div)
    if (!['svg', 'path'].includes(e.target.tagName)) return null;
    const classList = Array.from(e.target.classList);
    const provName = classList[classList.length - 1];
    if (!provName) return null;
    return {
      fullName: provinces[provName.slice(-2).toLowerCase()],
      shortName: provName.slice(-2).toLowerCase()
    }
  }

  handleClick(e) {
    const clickedProvince = this.getProvinceName(e);
    if (!clickedProvince) return;

    // if command or control is pressed Redirect in new window
    if ((e.ctrlKey || e.metaKey) && e.button === 0) {
      window.open(`/discover/${clickedProvince.shortName}`) ;
    } else if (e.which === 13 || e.button === 0) {
      this.props.history.push(`/discover/${clickedProvince.shortName}`)
    }
  }

  handleHover(e) {
    const hoveredProvince = this.getProvinceName(e);
    if (!hoveredProvince) {
        this.setState({ hoveredProvince: null })
    } else {
      this.setState({ hoveredProvince: hoveredProvince.fullName })
    }
  }

  render() {
    return (
      <React.Fragment>
        <h1>Discover Campgrounds in Canada</h1>
        <p>Explore all the campgrounds we have, from hiking the Rocky Mountains to
          the fields of Prince Edward Island. Click a province to see its campgrounds.</p>
        <div className='DiscoverSVGMap'
          onClick={this.handleClick}
          onMouseOver={this.handleHover}
          onMouseOut={() => this.setState({hoveredProvince: null})}
          onKeyDown={this.handleClick}
        >
          <CanadaMapSVG />
          <span className='DiscoverSVGMap__map-text bold flex-center'>
            Discover
            {this.state.hoveredProvince
              ?
              ` ${this.state.hoveredProvince} `
              :
              // Show scrolling list of provinces
              <div className='DiscoverSVGMap__scroll'>
                {Object.keys(provinces).map((prov, idx) => (
                  <span key={idx} className='DiscoverSVGMap__scroll-prov'>{prov}</span>
                ))
              }</div>
            }
            Campgrounds
          </span>
        </div>
      </React.Fragment>
    )
  }
}

export default DiscoverMap
