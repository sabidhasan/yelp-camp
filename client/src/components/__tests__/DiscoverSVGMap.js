import React from 'react'
import toJson from 'enzyme-to-json'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import DiscoverSVGMap from '../DiscoverSVGMap'

configure({ adapter: new Adapter() });

const getEventObject = (type, classList) => {
  return {
    target: {
      tagName: 'path',
      classList,
    },
    ...type
  }
}

describe('<DiscoverSVGMap />', () => {
  let component;
  let props;

  beforeEach(() => {
    props = {
      history: [],
    }
    component = shallow(<DiscoverSVGMap { ...props } />)
  })

  it('Component should be defined', () => {
    expect(component).toBeDefined();
  });

  it('Clicking a province navigates to that province', () => {
    const event = getEventObject({button: 0}, ['a', 'ab']);
    component.find('.DiscoverSVGMap').simulate('click', event);
    expect(props.history.pop()).toBe('/discover/ab');
  })

  it('Right clicking a province does not do anything', () => {
    const event = getEventObject({button: 2}, ['a', 'ab']);
    component.find('.DiscoverSVGMap').simulate('click', event);
    expect(props.history.length).toBe(0);
  })

  it('Clicking on a non-province does not do anything', () => {
    const event = getEventObject({button: 0}, []);
    component.find('.DiscoverSVGMap').simulate('click', event);
    expect(props.history.length).toBe(0);
  })

  it('On page load, the scrolling list is shown', () => {
    expect(component.find('.DiscoverSVGMap__scroll')).toHaveLength(1);
  })

  it('Mousing over a province updates shown text', () => {
    const event = getEventObject({}, ['a', 'b', 'bc']);
    component.find('.DiscoverSVGMap').simulate('mouseOver', event);
    expect(component.find('.DiscoverSVGMap__map-text').text()).toBe('Discover british columbia Campgrounds');
  })

  it('Mousing out of province shows scrolling list again', () => {
    component.setState({hoveredProvince: {}});
    component.find('.DiscoverSVGMap').simulate('mouseOut');
    expect(component.find('.DiscoverSVGMap__scroll')).toHaveLength(1);
  })

  it('Pressing Enter navigates to page if it exists', () => {
    const event = getEventObject({which: 13}, ['a', 'b', 'on']);
    component.find('.DiscoverSVGMap').simulate('keydown', event);
    expect(props.history.pop()).toBe('/discover/on');
  })

  it('Pressing non-Enter key does not navigate anywhere else', () => {
    const event = getEventObject({which: 14}, ['a', 'b', 'on']);
    component.find('.DiscoverSVGMap').simulate('keydown', event);
    expect(props.history.pop()).toBe(undefined);
  })

  it('DiscoverSVGMap component snapshot matches what is expected', () => {
    expect(toJson(component)).toMatchSnapshot();
  })
})
