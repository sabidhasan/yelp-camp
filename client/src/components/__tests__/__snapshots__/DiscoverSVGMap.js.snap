// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DiscoverSVGMap /> DiscoverSVGMap component snapshot matches what is expected 1`] = `
<Fragment>
  <h1>
    Discover Campgrounds in Canada
  </h1>
  <p>
    Explore all the campgrounds we have, from hiking the Rocky Mountains to the fields of Prince Edward Island. Click a province to see its campgrounds.
  </p>
  <div
    className="DiscoverSVGMap"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    <CanadaMapSVG />
    <span
      className="DiscoverSVGMap__map-text bold flex-center"
    >
      Discover
      <div
        className="DiscoverSVGMap__scroll"
      >
        <span
          className="DiscoverSVGMap__scroll-prov"
          key="0"
        >
          bc
        </span>
        <span
          className="DiscoverSVGMap__scroll-prov"
          key="1"
        >
          ab
        </span>
        <span
          className="DiscoverSVGMap__scroll-prov"
          key="2"
        >
          sk
        </span>
        <span
          className="DiscoverSVGMap__scroll-prov"
          key="3"
        >
          mb
        </span>
        <span
          className="DiscoverSVGMap__scroll-prov"
          key="4"
        >
          on
        </span>
        <span
          className="DiscoverSVGMap__scroll-prov"
          key="5"
        >
          qc
        </span>
        <span
          className="DiscoverSVGMap__scroll-prov"
          key="6"
        >
          pe
        </span>
        <span
          className="DiscoverSVGMap__scroll-prov"
          key="7"
        >
          nb
        </span>
        <span
          className="DiscoverSVGMap__scroll-prov"
          key="8"
        >
          ns
        </span>
        <span
          className="DiscoverSVGMap__scroll-prov"
          key="9"
        >
          nl
        </span>
        <span
          className="DiscoverSVGMap__scroll-prov"
          key="10"
        >
          yt
        </span>
        <span
          className="DiscoverSVGMap__scroll-prov"
          key="11"
        >
          nt
        </span>
        <span
          className="DiscoverSVGMap__scroll-prov"
          key="12"
        >
          nu
        </span>
      </div>
      Campgrounds
    </span>
  </div>
</Fragment>
`;
